### PR TITLE
Add direct tpm support for spire-agent

### PR DESCRIPTION
**Pull request by kfox1111**

### CODE REVIEW

File: charts/spire/charts/spire-agent/templates/configmap.yaml [hunk 1]
```diff
@@ -56,11 +56,27 @@ agent:
...
     k8s_psat:
       plugin_data:
         cluster: {{ include "spire-lib.cluster-name" . | quote }}
+  {{- $nodeAttestorUsed = add1 $nodeAttestorUsed }}
+  {{- end }}
+  {{- with .Values.nodeAttestor.tpmDirect }}
+  {{- if eq (.enabled | toString) "true" }}
+    tpm:
+      plugin_cmd: "/tpm/tpm_attestor_agent"
+      plugin_checksum: {{ .checksum }}
```
Comment (kfox1111) [inline charts/spire/charts/spire-agent/templates/configmap.yaml:72]:
```suggestion
      plugin_checksum: {{ .plugin.checksum }}
```

File: charts/spire/charts/spire-agent/templates/configmap.yaml [hunk 2]
```diff
@@ -56,11 +56,27 @@ agent:
   {{- end }}
   {{- end }}
 
+{{- $nodeAttestorUsed := 0 }}
```
Comment (kfox1111) [inline charts/spire/charts/spire-agent/templates/configmap.yaml:59]:
```suggestion
{{- $nodeAttestorUsed := add (len nodeAttestor) (len .Values.unsupportedBuiltInPlugins.nodeAttestor) }}
```

File: examples/tpm-direct/values.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+      enabled: false
+    tpmDirect:
+      enabled: true
+
+spire-server:
+  controllerManager:
+    # K8s labels have a 63 character limit. TPM hashes are 64 chars. So you need to label the node with two labels with half of the tpm's hash each. The 'node-restriction.kubernetes.io/' prefix is so that the
+    # nodes can't update the hash themselves, an important security constraint.
+    parentIDTemplate: 'spiffe://{{ .TrustDomain }}/spire/agent/tpm/{{ index .NodeMeta.Labels "node-restriction.kubernetes.io/tpm-pubhash" }}{{ index .NodeMeta.Labels "node-restriction.kubernetes.io/tpm-pubhash2" }}'
+    tpmDirect:
```
Comment (kfox1111) [inline examples/tpm-direct/values.yaml:13]:
```suggestion
  nodeAttestor:
    tpmDirect:
```

File: examples/tpm-direct/values.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    k8sPsat:
+      enabled: false
+    tpmDirect:
+      enabled: true
+
+spire-server:
+  controllerManager:
+    # K8s labels have a 63 character limit. TPM hashes are 64 chars. So you need to label the node with two labels with half of the tpm's hash each. The 'node-restriction.kubernetes.io/' prefix is so that the
+    # nodes can't update the hash themselves, an important security constraint.
+    parentIDTemplate: 'spiffe://{{ .TrustDomain }}/spire/agent/tpm/{{ index .NodeMeta.Labels "node-restriction.kubernetes.io/tpm-pubhash" }}{{ index .NodeMeta.Labels "node-restriction.kubernetes.io/tpm-pubhash2" }}'
```
Comment (marcofranssen) [inline examples/tpm-direct/values.yaml:12]: Should the chart check this `restriction.kubernetes.io` is part of the template as some sort of safetynet?

File: charts/spire/charts/spire-agent/values.yaml [hunk 5]
```diff
@@ -166,6 +166,37 @@ nodeAttestor:
   k8sPsat:
     ## @param nodeAttestor.k8sPsat.enabled Enable Psat k8s Node Attestor
     enabled: true
+  tpmDirect:
```
Comment (marcofranssen) [inline charts/spire/charts/spire-agent/values.yaml:169]: Maybe just name this. Trying to understand what Direct means in this context.

```suggestion
  tpm:
```

**Review by marcofranssen — COMMENTED** [review]
Small question and suggestion. Looks pretty neat.

File: charts/spire/charts/spire-agent/values.yaml (hunk 5 above)
Comment (kfox1111) [reply charts/spire/charts/spire-agent/values.yaml:169]: Marked as Direct to keep room for a second feature in the future. There is a second tpm mode that spire supports here: https://github.com/spiffe/spire/blob/main/doc/plugin_agent_nodeattestor_tpm_devid.md

Direct uses the tpm's private key hash directly for identification. This is the simplest way to use a tpm.

DeviceID requires an extra server and an out of band registration mechanism. But is the only way to attest some devices (Apple I think)

File: examples/tpm-direct/values.yaml (hunk 4 above)
Comment (kfox1111) [reply examples/tpm-direct/values.yaml:12]: I wish we could, but it is a restriction on the labels being manually added to the node by the end user. Its not done in the chart, so not something we can check in the chart.

k8s itself does enforce the restriction.If you try and load it all in one string, it complains that the label is too long.

Once the tpm plugin pr's merge, I'm going to provide documentation at https://github.com/spiffe/spiffe.io/pull/290

File: charts/spire/charts/spire-agent/values.yaml (hunk 5 above)
Comment (marcofranssen) [reply charts/spire/charts/spire-agent/values.yaml:169]: ok, thanks for explaining, makes sense to keep room for extension. Is `direct` a uniformal way of referring to this tpm approach? If so I'm fine. Just wanter to pick your brain on the naming convention.

File: examples/tpm-direct/values.yaml (hunk 4 above)
Comment (marcofranssen) [reply examples/tpm-direct/values.yaml:12]: Yes outside of the chart an enduser could do that indeed. But this value specifically we can control it right?

File: charts/spire/charts/spire-agent/values.yaml (hunk 5 above)
Comment (kfox1111) [reply charts/spire/charts/spire-agent/values.yaml:169]: I don't know of any standard naming around it. Also open to different names. Though naming things is hard.

File: examples/tpm-direct/values.yaml (hunk 4 above)
Comment (kfox1111) [reply examples/tpm-direct/values.yaml:12]: Oh. I think I may have been misunderstanding what you've been saying...  I thought you were talking about the character limit, not the parentIDTemplate itself.

That would be good to do if we could figure out a way to do it. I can't think of a way off hand to be able to do that, outside of writing our own gotl parser?

Comment (faisal-memon) [reply examples/tpm-direct/values.yaml:12]: @marcofranssen Are you asking just to check if the string `restriction.kubernetes.io` is present?

File: charts/spire/charts/spire-agent/values.yaml (hunk 5 above)
Comment (marcofranssen) [reply charts/spire/charts/spire-agent/values.yaml:171]: Yup agree. I'm not very into tpm stuff so just asked my 101 questions here on the PR. I guess it is good to go. @faisal-memon feel free to approve and merge if you don't see any other hurdles.

File: examples/tpm-direct/values.yaml (hunk 4 above)
Comment (marcofranssen) [reply examples/tpm-direct/values.yaml:12]: it was about the char limit indeed, but seeing Kevins response that is not possible prior to parsing the template I guess. So no worries, I'm fine merging like this. Just tried to see if we could failsafe the user.

File: charts/spire/charts/spire-agent/templates/daemonset.yaml [hunk 6]
```diff
@@ -41,13 +46,55 @@ spec:
...
+            - sh
+            - -c
+            - |
+              TPM=`/app/get_tpm_pubhash 2> /dev/null`
+              if [ $? -eq 0 ]; then
+                echo "TPM Fingerprint: $TPM"
+                (echo -n '<1>'; echo "TPM Fingerprint: $TPM") | tee /dev/kmsg
+              else
+                /app/get_tpm_pubhash
+              fi
```
Comment (faisal-memon) [inline charts/spire/charts/spire-agent/templates/daemonset.yaml:65]: Can you add a comment here explaining what this snippet does?

File: charts/spire/charts/spire-agent/templates/daemonset.yaml [hunk 7]
```diff
@@ -41,13 +46,55 @@ spec:
...
+        {{- if .Values.nodeAttestor.tpmDirect.pubHash.enabled }}
+        - name: fingerprint-tpm
+          securityContext:
+            {{- $mainSecurityContext | toYaml | nindent 12 }}
+          image: {{ template "spire-lib.image" (dict "appVersion" $.Chart.AppVersion "image" .Values.nodeAttestor.tpmDirect.pubHash.image "global" .Values.global) }}
+          command:
+            - sh
+            - -c
+            - |
+              TPM=`/app/get_tpm_pubhash 2> /dev/null`
```
Comment (kfox1111) [inline charts/spire/charts/spire-agent/templates/daemonset.yaml:59]:
```suggestion
              # Fetch the TPM's fingerprint from the TPM and output it on the Computers Console. This enables secure bootstrapping of the Node with the Server.
              TPM=`/app/get_tpm_pubhash 2> /dev/null`
```

**Review by faisal-memon — APPROVED** [review]

File: charts/spire/charts/spire-agent/values.yaml (hunk 5 above)
Comment (faisal-memon) [reply charts/spire/charts/spire-agent/values.yaml:174]: Naming looks ok to me. Keeping room for the second tpm plugin makes sense

**Pull request merged**